### PR TITLE
Fixed Mini BO avatar images not deleting when updating images

### PR DESCRIPTION
**Pull request by db306**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | When updating an object that inherits the ObjectModel with a new image, the cached thumbnail generated by the HelperList does not update as it doesn't get deleted properly.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes [BOOM-5503](http://forge.prestashop.com/browse/BOOM-5503) ticket
| How to test?  | Requires a custom implementation of an ObjectModel that uses the HelperList Class and displays an image.

When displaying images of a custom ObjectModel in a list using the **HelperList** Class within the BO, it generates a thumbnail of 45px that is later on displayed on the table. This image generation is triggered on : `classes/helper/HelperList.php` on line **312** : 

```
$this->_list[$index][$key] = ImageManager::thumbnail($path_to_image, $this->table.'_mini_'.$item_id.'_'.$this->context->shop->id.'.'.$this->imageType, 45, $this->imageType);
```
When updating an image of a custom ObjectModel implementation, it triggers the function deleteImage from the ObjectModel class that deletes existing images:

**/classes/ObjectModel.php** lines **1753** to **1769**

```
if ($this->image_dir) {
    if (file_exists($this->image_dir.$this->id.'.'.$this->image_format)
        && !unlink($this->image_dir.$this->id.'.'.$this->image_format)) {
        return false;
    }
}
if (file_exists(_PS_TMP_IMG_DIR_.$this->def['table'].'_'.$this->id.'.'.$this->image_format)
    && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_'.$this->id.'.'.$this->image_format)) {
    return false;
}
if (file_exists(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'.'.$this->image_format)
    && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'.'.$this->image_format)) {
    return false;
}
```

The former piece of code will create an image in a path structure such as `/img/tmp/customtable_mini_15_1.jpg` whereas the equivalent example on the later piece of code will delete an image in a uri `/img/tmp/customtable_mini_15.jpg` given that these urls are not the same, when updating an image it will not delete its "cached" thumbnail.

Concerning the PR I have added this piece of code that checks whether the ObjectModel supports Multishop and then iterates through all shop ids or a single shop id depending on the current view. Then deletes existing thumbnail images.

Another way, and **probably a best approach** would be to change the HelperList thumbnail generation to stop using the shop_id as a variable in its path url. I assumed this had a reason to be here and hence I have changed the ObjectModel function instead. Let me know and I can modify my PR to change the thumbnail url instead :

This :
```
ImageManager::thumbnail($path_to_image, $this->table.'_mini_'.$item_id.'.'.$this->imageType, 45, $this->imageType);
```

instead of currently this:
```
ImageManager::thumbnail($path_to_image, $this->table.'_mini_'.$item_id.'_'.$this->context->shop->id.'.'.$this->imageType, 45, $this->imageType);
```

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9045)

### CODE REVIEW

**Comment by db306** [comment]
`description` isn't empty. Or is it?

File: classes/ObjectModel.php [hunk 1]
```diff
@@ -1767,6 +1767,20 @@ public function deleteImage($force_delete = false)
                 && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'.'.$this->image_format)) {
                 return false;
             }
+            if($this->isMultishop()){
```
Comment (PierreRambaud) [inline classes/ObjectModel.php]: Will be easier to read and remove duplicates :)

```php
if ($this->isMultishop()) {
    $ids = $this->id_shop;
    if (!is_array($ids)) {
        $ids = [$ids];
    }

    foreach ($ids as $idShop) {
        $filePath = _PS_TMP_IMG_DIR_ .
                  $this->def['table'] .
                  '_mini_' .
                  $this->id .
                  '_' .
                  $idShop .
                  '.' .
                  $this->image_format;
        if (file_exists($filePath) && !unlink($filePath) {
            return false;
        }
    }
}
```

Comment (db306) [reply classes/ObjectModel.php]: 👍

File: classes/ObjectModel.php [hunk 2]
```diff
@@ -1767,15 +1767,15 @@ public function deleteImage($force_delete = false)
                 && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'.'.$this->image_format)) {
                 return false;
             }
-            if($this->isMultishop()){
+            if ($this->isMultishop()){
 
                 $ids_shop = $this->id_shop;
```
Comment (PierreRambaud) [inline classes/ObjectModel.php]: Could be better with camelcase instead of underscore variable :)

File: classes/ObjectModel.php [hunk 3]
```diff
@@ -1767,15 +1767,15 @@ public function deleteImage($force_delete = false)
...
-                if(!is_array($this->id_shop)){
+                if (!is_array($this->id_shop)){
                     $ids_shop = [$this->id_shop];
                 }
-                foreach ($ids_shop as $id_shop){
-                    if(file_exists(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'_'.$id_shop.'.'.$this->image_format)
-                        && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'_'.$id_shop.'.'.$this->image_format)){
+                foreach ($ids_shop as $id_shop) {
+                    if (file_exists(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'_'.$id_shop.'.'.$this->image_format)
```
Comment (PierreRambaud) [inline classes/ObjectModel.php]: Can you please put file path into a variable to remove duplicate concatenation?

File: classes/ObjectModel.php (hunk 2 above)
Comment (db306) [reply classes/ObjectModel.php]: sure

File: classes/ObjectModel.php (hunk 3 above)
Comment (db306) [reply classes/ObjectModel.php]: I would heave done it that way too so I cannot agree more, but the whole function is built with double concatenation. I'll change them all and give them meaningful variable names.

Comment (PierreRambaud) [reply classes/ObjectModel.php]: Nice, really easier to read now!

**Comment by jolelievre** [comment]
seems good to me!
@PierreRambaud ?

**Review by PierreRambaud — DISMISSED** [review]

**Comment by PierreRambaud** [comment]
Thanks @jolelievre forgot this one.
@db306 Can you send us a module or a script to test it? :)

**Comment by db306** [comment]
Thanks, I’ll look if I can make some tests. I’m leaving on vacations at the moment so it will have to wait til i get back.

**Comment by db306** [comment]
Hello, it says the branch has conflicts that must be resolved ? I don't understand what is conflicting here

**Comment by PierreRambaud** [comment]
In fact the conflict is coming when it tries to merge your branch into develop branch.
You can use `git rebase` command or `git pull`

**Comment by PierreRambaud** [comment]
Hum, look like something went wrong ^^
`Files changed 4,256 +216,018 −237,677`

**Comment by PierreRambaud** [comment]
Can you please rebase your branch?

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Waiting for rebase

**Comment by marionf** [comment]
Hello @db306 

Since we have had no news for more than 30 days, I close this PR
Feel free to open a new one if needed

**Pull request closed without merging**